test/common/crypto/utility_test.cc

```diff
@@ -159,30 +212,816 @@ TEST(UtilityTest, TestVerifySignature) {
+
+  // Test with corrupted key that might cause EVP_DigestVerifyInit to fail
+  auto corrupted_key = std::make_unique<PKeyObject>();
+  // Create a partially valid but corrupted EVP_PKEY that might cause init failures
+  EVP_PKEY* corrupted_pkey = EVP_PKEY_new();
+  corrupted_key->setEVP_PKEY(corrupted_pkey);
+
+  std::vector<uint8_t> test_data = {'t', 'e', 's', 't'};
+  std::vector<uint8_t> dummy_sig = {1, 2, 3, 4};
+
+  auto verify_result =
+      UtilitySingleton::get().verifySignature("sha256", *corrupted_key, dummy_sig, test_data);
+  EXPECT_FALSE(verify_result.ok())
+      << "Corrupted key should fail verification: " << verify_result.message();
+}
+
+TEST(UtilityTest, TestHelperFunctionsCoverage) {
+  // Test helper functions directly to improve coverage
+  auto impl = std::make_unique<UtilityImpl>();
+
+  // Test empty key string (exercises BIO allocation with edge case)
+  std::string empty_key;
+  auto empty_public_pem = impl->importPublicKeyPEM(empty_key);
+  EXPECT_EQ(nullptr, empty_public_pem->getEVP_PKEY()) << "Empty PEM key should fail";
+
+  auto empty_private_pem = impl->importPrivateKeyPEM(empty_key);
+  EXPECT_EQ(nullptr, empty_private_pem->getEVP_PKEY()) << "Empty PEM private key should fail";
+
+  // Test DER functions directly with various invalid inputs
+  std::vector<uint8_t> invalid_der = {0x30, 0x82}; // Truncated DER
+  auto der_public = impl->importPublicKeyDER(invalid_der);
+  EXPECT_EQ(nullptr, der_public->getEVP_PKEY()) << "Invalid DER key should fail";
+
+  auto der_private = impl->importPrivateKeyDER(invalid_der);
+  EXPECT_EQ(nullptr, der_private->getEVP_PKEY()) << "Invalid DER private key should fail";
+}
+
+TEST(UtilityTest, TestImportKeyErrorPaths) {
+  // Test additional error paths in key import functions
+  auto impl = std::make_unique<UtilityImpl>();
+
+  // Test with null data pointer (edge case)
+  std::string null_data(100, '\0'); // 100 null characters
+  // This tests the BIO_new_mem_buf with null data
+  auto null_public_pem = impl->importPublicKeyPEM(null_data);
+  EXPECT_EQ(nullptr, null_public_pem->getEVP_PKEY()) << "Null data should fail PEM import";
+
+  auto null_private_pem = impl->importPrivateKeyPEM(null_data);
+  EXPECT_EQ(nullptr, null_private_pem->getEVP_PKEY()) << "Null data should fail private PEM import";
+
+  // Test with data that looks like PEM but has invalid structure
+  std::string malformed_pem = "-----BEGIN PUBLIC KEY-----\n"
+                              "This is not valid base64 content\n"
+                              "-----END PUBLIC KEY-----";
+
+  auto malformed_public = impl->importPublicKeyPEM(malformed_pem);
+  EXPECT_EQ(nullptr, malformed_public->getEVP_PKEY()) << "Malformed PEM should fail";
+
+  auto malformed_private = impl->importPrivateKeyPEM(malformed_pem);
+  EXPECT_EQ(nullptr, malformed_private->getEVP_PKEY()) << "Malformed private PEM should fail";
+}
+
+TEST(UtilityTest, TestEdgeCasesAndMissingCoverage) {
```

Comment:
This isn't a good test name. Maybe have separate tests for some of these, so you can describe clearly what it's actually testing.